### ISSUE TITLE
Don't try to delete/process date archives if settings prevent it

### ISSUE THREAD

**Issue by brutaldev** [NLog/NLog#1147]
I'm getting the following exception when using archive date patterns such as `"_yyyyMMdd_HH"`

```
NLog.NLogRuntimeException: Exception occurred in NLog ---> System.ArgumentOutOfRangeException: Index and length must refer to a location within the string.
Parameter name: length
   at System.String.Substring(Int32 startIndex, Int32 length)
   at NLog.Targets.FileTarget.DeleteOldDateArchives(String pattern)
   at NLog.Targets.FileTarget.Write(LogEventInfo logEvent)
   at NLog.Targets.Target.Write(AsyncLogEventInfo logEvent)`
```

This goes through a significant amount of code starting at line [739 of FileTarget.cs](https://github.com/NLog/NLog/blob/master/src/NLog/Targets/FileTarget.cs#L739) which breaks before reaching 'EnsureArchiveCount' which is what actually deletes the file.

'EnsureArchiveCount' short-circuits immediately with `if (this.MaxArchiveFiles <= 0) return;`

I believe that it should not go into the conditional statement in line [739](https://github.com/NLog/NLog/blob/master/src/NLog/Targets/FileTarget.cs#L739) if `MaxArchiveFiles <= 0` to avoid trying to delete anything when you clearly don't need to.

**Comment by 304NotModified** [comment]
Hey! When do you get this exception? Can you create an unit test for this?

**Comment by brutaldev** [comment]
Pretty easy to get it to happen, using the Configuration API with a couple lines of code already so I can turn that into a unit test for you.

I've got a custom build that fixes the issue by avoid that whole conditional block as an optimization. It will still happen though if you set `MaxArchiveFiles > 0` so it will be worth fixing the real underlying issue as well.

**Comment by brutaldev** [comment]
Added a fix and an optimization around this issue in pull request #1149.

This is potentially a dangerous issue since files that aren't actually archive files may inadvertently be deleted because they match the pattern.

I found this because I have a logs directory that is used by many applications. There were two log files called `Debug_App1.log` and `Debug_App2.log` respectively. In this scenario it's very likely that archives for App1 could be deleted when App2 is running even though App1 is configured to never delete archives.

So although this pull request fixes the crash, it doesn't solve the potential issue of deleting files that aren't archive files. To be safe for now I use different directories for each application's archives so they don't get included in the directory scan.
